lib/posix/pipe.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <posix/unistd.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include <sys/fdtable.h>
+
+struct pipe_fd_object {
+	struct stat stat;
+	/* counterpart file descriptor */
+	int other;
+};
+
+/* read side of pipe */
+struct pipe_fd_robject {
+	struct pipe_fd_object obj;
+};
+
+/* write side of pipe */
+struct pipe_fd_wobject {
+	struct pipe_fd_object obj;
+	struct k_pipe pipe;
+	unsigned char buf[CONFIG_POSIX_PIPE_SIZE];
+};
+
+static ssize_t pipe_read(void *obj, void *buf, size_t sz)
+{
+	errno = ENOSYS;
+	return -1;
+}
+
+static ssize_t pipe_write(void *obj, const void *buf, size_t sz)
+{
+	errno = ENOSYS;
+	return -1;
+}
+
+static int pipe_ioctl(void *obj, unsigned int request, va_list args)
+{
+	errno = ENOSYS;
+	return -1;
+}
+
+static const struct fd_op_vtable pipe_fd_op_vtable = {
+	.read = pipe_read,
+	.write = pipe_write,
+	.ioctl = pipe_ioctl,
+};
+
+static int pipe_fd_alloc(bool is_read_end, struct pipe_fd_object **obj)
+{
+
+	const size_t size = is_read_end ? sizeof(struct pipe_fd_robject) : sizeof(struct pipe_fd_wobject);
+
+	int res;
+	int fd;
+	void *o;
+
+	/* assert obj is non-null */
+
+	/* TODO: Check per-process fd limit and return EMFILE if not ok */
+
+	res = z_alloc_fd(obj, &pipe_fd_op_vtable);
+	if (res == -1) {
+		errno = ENFILE;
+		res = -1;
+		goto out;
+	}
+	fd = res;
+
+	o = k_malloc(size);
+	if (o == NULL) {
+		errno = ENFILE;
+		res = -1;
+		goto free_fd;
+	}
+	memset(o, 0, size);
+
+	if (is_read_end) {
+		*obj = & ((struct pipe_fd_robject *)o)->obj;
+		(*obj)->stat.st_mode = S_IRUSR;
+	} else {
+		struct pipe_fd_wobject *wobj = (struct pipe_fd_wobject *)o;
+		*obj = & wobj->obj;
+		(*obj)->stat.st_mode = S_IWUSR;
+		if (CONFIG_POSIX_PIPE_SIZE == 0) {
+			k_pipe_init(&wobj->pipe, NULL, 0);
+		} else {
+			k_pipe_init(&wobj->pipe, wobj->buf, sizeof(wobj->buf));
+		}
+	}
+	(*obj)->stat.st_mode = _IFIFO;
+
+	res = fd;
+	goto out;
+
+free_fd:
+	z_free_fd(fd);
+
+out:
+	return res;
+}
+
+static void pipe_join(int filedes[2], struct pipe_fd_object *obj[2])
```

Comment:
I was trying to think of a plumbing analogy, but yeah, either way works for me

Reply:
pipe_couple?

Reply:
It's just "join" is a term in (computational) thread terminology, effectively meaning "terminate" (cf. pthread_join() ). Avoiding reusing such terms to avoid confusion would be beneficial imho.

Reply:
Right. It turns out that "join" is also used for plumbing. I think I'll just put it in `pipe_fd_alloc()`